mautic/mautic
# Import on background

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/183
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/1642, partly https://github.com/mautic/mautic/issues/3906, hopefully https://github.com/mautic/mautic/issues/2630
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
The contact import is taking forever for bigger CSV files and it's unstable because it's done via AJAX HTTP requests which often has a short time limit. This PR changes the process to be handled as a background (CLI) job. The whole CSV will be processed in one run. This PR also adds:

- [x] An entity which holds information about every import
- [x] Permissions who can import.
- [x] Log entry into the user activities
- [x] Log entry into the contact activities
- [x] Ability to pause import when it runs
- [x] Parallel command control - do not allow to run too much imports in parallel
- [x] Notify use after import is finished
- [x] UI for past/current background imports
- [x] Create an option to decide whether to import in the browser or in the background based on the CSV row count. 
- [x] Document the 2-button configuration
- [x] Try to speed it up.
- [ ] Catch the validation exceptions and show a generic error message to the user for database exceptions, log the detailed error though.

#### Steps to test this PR:
0. Take a look at the linked documentation to be aware of what's new. 
0.1. During following steps, try to configure the import differently every time. Different default values like owner, segment, tags and different core fields like date created.
0.2 Apply this PR.
0.3 Clear the cache.
0.4 Run `app/console doctrine:migrations:migrate`
1. Try to import CSV with contacts as you are used to - inside the browser. Everything should work, the contacts should be created/updated accordingly.
2.  Try to import also via background job. When you configure the import for that, run command `app/console mautic:import`. It will show a progress bar. When the import is completed again see if all the contacts were created/updated with correct values.
3. Try to import a bigger CSV in the background and try to unpublish in the process. The import will stop. When you publish it again and start the command again, it will continue where it left off.
4. If you can try to import some massive CSV via the background job and see what it can handle.
5. There's a new option in the global config. Go to _Contact Settings_ tab and set the _Automatically import in the background if the CSV has more rows than defined_ option to 100 for example. Then try to import a CSV file with less than 100 rows again. You'll see there is only one _Import_ button. Hit it and it will start the import in the browser. Then try to import a CSV file with more than 100 rows and it will queue it for background processing.
